benchmarks/polys.py

```diff
@@ -1,17 +1,125 @@
+        return f, g, fp, gp, fpe, gpe
+
+
+class _SparseGCDHighDegree:
+    def generate(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        D = 1 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)])
+        f = D * (-2 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)]))
+        g = D * (2 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)]))
+        fp, gp = Poly(f, self.x, *(self.y)[:n]), Poly(g, self.x, *(self.y)[:n])
+        fpe, gpe = self.R.from_sympy(f), self.R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe
+
+
+class _QuadraticNonMonicGCD:
+    def generate(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        D = 1 + self.x ** 2 * (self.y)[0] ** 2 + sum([(self.y)[i] ** 2 for i in range(1, n)])
+        f = D * (-1 + self.x ** 2 - (self.y)[0] ** 2 + sum([(self.y)[i] ** 2 for i in range(1, n)]))
+        g = D * (2 + self.x * (self.y)[0] + sum((self.y)[1:n])) ** 2
+        fp, gp = Poly(f, self.x, *(self.y)[:n]), Poly(g, self.x, *(self.y)[:n])
+        fpe, gpe = self.R.from_sympy(f), self.R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe
+
+
+class _SparseNonMonicQuadratic:
+    def generate(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        D = -1 + self.x * prod((self.y)[:n])
+        f = D * (3 + self.x * prod((self.y)[:n]))
+        g = D * (-3 + self.x * prod((self.y)[:n]))
+        fp, gp = Poly(f, self.x, *(self.y)[:n]), Poly(g, self.x, *(self.y)[:n])
+        fpe, gpe = self.R.from_sympy(f), self.R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe
+
+
+class _TimePREM:
+
+    def setup(self, method, n):
+        (self.f, self.g, self.fp, self.gp, self.fpe, self.gpe) = self.generate(n)
+        self.values = {}
+        self.x = symbols("x")
+
+        if method == 'prem':
+            self.func = lambda: prem(self.f, self.g, self.x)
+
+        elif method == 'Poly_prem':
+            self.func = lambda: self.fp.prem(self.gp)
+
+        elif method == 'PolyElement_prem':
+            self.func = lambda: self.fpe.prem(self.gpe)
+
+    def time_prem_methods(self, method, n):
+        self.values[str(method)] = self.func()
+
+    def teardown(self, method, n):
+            expected_res = {
+                'prem': self.values['prem'],
+                'Poly_prem': self.values['Poly_prem'],
+                'PolyElement_prem': self.values['PolyElement_prem']
+            }
+
+            computed_res = self.values[method]
+            expected_res = expected_res[method]
+
+            if (computed_res - expected_res).simplify() != 0:
+                raise ValueError("Incorrect result, invalid timing:"
+                                " %s != %s" % (expected_res, computed_res))
```

Comment:
Is the right way to implement the `teardown` function?

Reply:
No because this is comparing the computed value with itself.

We either need to have the correct answer stored or we need to check some invariant of the answer. For prem one invariant is:
```python
In [24]: f = x**2 + y**2 - 1

In [25]: g = x + 3*y

In [26]: p = prem(f, g, x)

In [27]: rem(f*LC(g, x)**degree(g, x), g, x) == p
Out[27]: True
```

Reply:
I think it should be `rem(f*LC(g, x)**(degree(f, x) - degree(g, x) + 1), g, x)` 
Also for  `LC` there are different values for the rings and domains
 ```python
>>> x, y, z = symbols("x y z")
>>> R = ZZ[x, y, z]
>>> gpe = 2*x*y + 2*x*z
>>> gp  = R.from_sympy(gpe)
>>> gp.LC
2
``` 
```python
>>> x, y, z = symbols('x y z')
>>> gpe = 2*x*y + 2*x*z
>>> LC(gpe,  x)
2*y + 2*z
```

The function which can give the desired result is `coeff_wrt(f, x, deg)` which is implemented in https://github.com/sympy/sympy/pull/25278/files

Reply:
> For the failed cases there might be a problem with `LC`

You can convert from one form to another and back if needed e.g.:
```python
In [540]: R(gp.as_expr().as_poly(x).LC())
Out[540]: 2*y + 2*z
```